File: backend/models/campaign.py
"""
Campaign model for ad management
"""

from sqlalchemy import Column, Integer, String, DateTime, Boolean, Date, Text, ForeignKey, Enum as SQLEnum
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from backend.database import Base
import enum


class ApprovalStatus(str, enum.Enum):
    """Approval status enumeration"""
    PENDING = "PENDING"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    NOT_REQUIRED = "NOT_REQUIRED"


class RateType(str, enum.Enum):
    """Rate type enumeration"""
    ROS = "ROS"  # Run of Schedule
    DAYPART = "DAYPART"  # Daypart-specific
    PROGRAM = "PROGRAM"  # Program-specific
    FIXED_TIME = "FIXED_TIME"  # Fixed time slot


class Campaign(Base):
    """Campaign model for ad management - extended with order/contract fields"""
    __tablename__ = "campaigns"

    id = Column(Integer, primary_key=True, index=True)
    advertiser = Column(String(255), nullable=False, index=True)  # Keep for backward compatibility
    start_date = Column(Date, nullable=False, index=True)
    end_date = Column(Date, nullable=False, index=True)
    priority = Column(Integer, default=1, index=True)
    file_url = Column(Text)
    active = Column(Boolean, default=True, index=True)
    
    # New order/contract fields
    order_number = Column(String(50), unique=True, nullable=True, index=True)
    contract_number = Column(String(50), nullable=True, index=True)
    insertion_order_url = Column(Text, nullable=True)
    spot_lengths = Column(JSONB)  # Array of spot lengths in seconds
    rate_type = Column(SQLEnum(RateType), nullable=True)
    rates = Column(JSONB)  # Rate structure based on rate_type
    scripts = Column(JSONB)  # Array of script/copy references
    copy_instructions = Column(Text, nullable=True)
    traffic_restrictions = Column(JSONB)  # Restrictions and requirements
    approval_status = Column(SQLEnum(ApprovalStatus), nullable=True, default=ApprovalStatus.NOT_REQUIRED)
    approved_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    approved_at = Column(DateTime(timezone=True), nullable=True)
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    # Relationships
    orders = relationship("Order", back_populates="campaign")
    spots = relationship("Spot", back_populates="campaign")
    makegoods = relationship("Makegood", back_populates="campaign")
    invoices = relationship("Invoice", back_populates="campaign")
    approver_user = relationship("User", foreign_keys=[approved_by])

    def __repr__(self):
        return f"<Campaign(advertiser='{self.advertiser}', priority={self.priority})>"

File: archive/librelog-web-api-original/src/main/java/com/onelpro/librelog/models/Campaign.java
package com.onelpro.librelog.models;

import com.onelpro.librelog.enums.ApprovalStatus;
import com.onelpro.librelog.enums.RateType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

/**
 * Entity representing a campaign in the system.
 */
@Entity
@Table(name = "campaigns", indexes = {
        @Index(name = "ix_campaigns_id", columnList = "id"),
        @Index(name = "ix_campaigns_advertiser", columnList = "advertiser"),
        @Index(name = "ix_campaigns_start_date", columnList = "start_date"),
        @Index(name = "ix_campaigns_end_date", columnList = "end_date"),
        @Index(name = "ix_campaigns_priority", columnList = "priority"),
        @Index(name = "ix_campaigns_active", columnList = "active"),
        @Index(name = "ix_campaigns_order_number", columnList = "order_number", unique = true),
        @Index(name = "ix_campaigns_contract_number", columnList = "contract_number")
})
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Campaign {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "advertiser", nullable = false, length = 255)
    private String advertiser;

    @Column(name = "start_date", nullable = false)
    private LocalDate startDate;

    @Column(name = "end_date", nullable = false)
    private LocalDate endDate;

    @Column(name = "priority")
    private Integer priority;

    @Column(name = "file_url", columnDefinition = "TEXT")
    private String fileUrl;

    @Column(name = "active")
    private Boolean active;

    @Column(name = "order_number", length = 50, unique = true)
    private String orderNumber;

    @Column(name = "contract_number", length = 50)
    private String contractNumber;

    @Column(name = "insertion_order_url", columnDefinition = "TEXT")
    private String insertionOrderUrl;

    @Column(name = "spot_lengths", columnDefinition = "JSONB")
    private String spotLengths;

    @Enumerated(EnumType.STRING)
    @Column(name = "rate_type", columnDefinition = "ratetype")
    private RateType rateType;

    @Column(name = "rates", columnDefinition = "JSONB")
    private String rates;

    @Column(name = "scripts", columnDefinition = "JSONB")
    private String scripts;

    @Column(name = "copy_instructions", columnDefinition = "TEXT")
    private String copyInstructions;

    @Column(name = "traffic_restrictions", columnDefinition = "JSONB")
    private String trafficRestrictions;

    @Enumerated(EnumType.STRING)
    @Column(name = "approval_status", columnDefinition = "approvalstatus")
    @Builder.Default
    private ApprovalStatus approvalStatus = ApprovalStatus.NOT_REQUIRED;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "approved_by")
    private User approvedBy;

    @Column(name = "approved_at")
    private Instant approvedAt;

    @Column(name = "created_at")
    private Instant createdAt;

    @Column(name = "updated_at")
    private Instant updatedAt;
}

